keras/backend/tensorflow/numpy.py

```diff
@@ -688,26 +695,64 @@ def linspace(
 
 @sparse.densifying_unary(-tfnp.inf)
 def log(x):
-    return tfnp.log(x)
+    x = convert_to_tensor(x)
+    dtype = (
+        config.floatx()
+        if standardize_dtype(x.dtype) == "int64"
+        else dtypes.result_type(x.dtype, float)
+    )
+    x = tf.cast(x, dtype)
+    # TODO: tfnp.log incorrectly promote bfloat16 to float64
+    return tf.cast(tfnp.log(x), dtype)
 
 
 @sparse.densifying_unary(-tfnp.inf)
 def log10(x):
-    return tfnp.log10(x)
+    x = convert_to_tensor(x)
+    dtype = (
+        config.floatx()
+        if standardize_dtype(x.dtype) == "int64"
+        else dtypes.result_type(x.dtype, float)
+    )
+    x = tf.cast(x, dtype)
+    # TODO: tfnp.log10 incorrectly promote bfloat16 to float64
+    return tf.cast(tfnp.log10(x), dtype)
 
 
 @sparse.elementwise_unary
 def log1p(x):
-    return tfnp.log1p(x)
+    x = convert_to_tensor(x)
+    dtype = (
+        config.floatx()
+        if standardize_dtype(x.dtype) == "int64"
+        else dtypes.result_type(x.dtype, float)
+    )
+    x = tf.cast(x, dtype)
+    # TODO: tfnp.log1p incorrectly promote bfloat16 to float64
+    return tf.cast(tfnp.log1p(x), dtype)
 
 
 @sparse.densifying_unary(-tfnp.inf)
 def log2(x):
-    return tfnp.log2(x)
+    x = convert_to_tensor(x)
+    dtype = (
+        config.floatx()
+        if standardize_dtype(x.dtype) == "int64"
+        else dtypes.result_type(x.dtype, float)
+    )
+    x = tf.cast(x, dtype)
+    # TODO: tfnp.log10 incorrectly promote bfloat16 to float64
+    return tf.cast(tfnp.log2(x), dtype)
 
 
 def logaddexp(x1, x2):
-    return tfnp.logaddexp(x1, x2)
+    x1 = convert_to_tensor(x1)
+    x2 = convert_to_tensor(x2)
+    dtype = dtypes.result_type(x1.dtype, x2.dtype, float)
+    x1 = tf.cast(x1, dtype)
+    x2 = tf.cast(x2, dtype)
+    # TODO: tfnp.logaddexp incorrectly promote bfloat16 to float64
```

Comment:
Is it intentional? `log` operations in some cases provide more accurate results with double precision. I am not sure though if it's the same reason here

Reply:
Hi @AakashKumarNain 
I'm not sure if I understand your question correctly.

We need the extra `cast` because `tfnp.log*` will accidentally upcast the tensor from bfloat16 to float64. This behavior happens ONLY at bfloat16.

All the other backends will maintain the original dtype.

```python
import jax.numpy as jnp
import numpy as np
import torch
import tensorflow as tf


tfnp = tf.experimental.numpy
tfnp.experimental_enable_numpy_behavior()

# jax
x = jnp.array(10, dtype="bfloat16")
y = jnp.log(x)
print(y, y.dtype)

# tensorflow
x = tf.constant(10, dtype="bfloat16")
y = tfnp.log(x)
print(y, y.dtype)

# torch
x = torch.tensor(10, dtype=torch.bfloat16)
y = torch.log(x)
print(y, y.dtype)

# np (using higher precision for checking)
x = np.array(10, dtype="float32")
y = np.log(x)
print(y)

```

```bash
2.29688 bfloat16
tf.Tensor(2.302585092994046, shape=(), dtype=float64) <dtype: 'float64'>
tensor(2.2969, dtype=torch.bfloat16) torch.bfloat16
2.3025851

# we can see that tensorflow has a more accurate output compared to jax and torch
# but we need to cast the output to the original dtype
```

Reply:
This is likely a tfnp bug.